examples/pwm_advanced.rs

```diff
@@ -0,0 +1,411 @@
+//! PA0  - TIM2 CH1   - active high PWM 50Hz   75% duty right aligned
+//! PA1  - TIM2 CH2   - active high PWM 50Hz   50% duty right aligned
+//! PA2  - TIM2 CH3   - active high PWM 50Hz   25% duty right aligned
+//! PB11 - TIM2 CH4   - active low  PWM 50Hz   75% duty right aligned
+//! PB5  - TIM3 CH2   - active high PWM 1kHz   33% duty center aligned
+//! PB0  - TIM3 CH3   - active high PWM 1kHz   67% duty center aligned
+//! PD15 - TIM4 CH4   - active low  PWM 2kHz   40% duty left aligned
+//! PA3  - TIM5 CH4   - active high PWM 2.5kHz 80% duty center aligned
+//! PG3  - TIM8 BKIN2 - active high fault input (3.3V is fault)
+//! PC6  - TIM8 CH1   - active high PWM 4kHz   50% duty right aligned
+//! PA5  - TIM8 CH1N  - active high complementary 2us deadtime
+//! PC7  - TIM8 CH2   - active high PWM 4kHz   10% duty right aligned
+//! PC8  - TIM8 CH3   - active high PWM 4kHz   25% duty right aligned
+//! PB1  - TIM8 CH3N  - active high complementary 2us deadtime
+//! PC9  - TIM8 CH4   - active low  PWM 4kHz   75% duty right aligned
+//! PB14 (LD3) - TIM12 CH1 - active low PWM 5kHz 25% duty left aligned
+//! PB15 - TIM12 CH2  - active high PWM 5kHz   40% duty left aligned
+//! PF8  - TIM13 CH1  - active low  PWM 6kHz   30% duty left aligned
+//! PA7  - TIM14 CH1  - active high PWM 7kHz   20% duty left aligned
+//! PE3  - TIM15 BKIN - active low  fault input (GND is fault)
+//! PE5  - TIM15 CH1  - active low  PWM 500kHz 25% duty left aligned
+//! PE4  - TIM15 CH1N - active low  complementary no deadtime
+//! PE6  - TIM15 CH2  - active high PWM 500kHz 50% duty left aligned
+//! PF10 - TIM16 BKIN - active high fault input (3.3V is fault)
+//! PF6  - TIM16 CH1  - active low  PWM 500kHz 25% duty left aligned
+//! PF7  - TIM17 CH1  - active high PWM 500kHz 50% duty left aligned
+//! PF9  - TIM17 CH1N - active high complementary 250ns deadtime
+//!
+//!
+//! If you start running with PE15 connected to VDD, you should see PWM on PE8-PE14.
+//! If you momentarily pull PE15 low, LD2 will turn on and you will get no PWM on PE8-PE14, even if you pull PE15 high again
+//! If PE15 is high and you press the USER button, TIM1 PWM will resume on PE8-PE14
+
+#![deny(warnings)]
+#![no_main]
+#![no_std]
+
+use cortex_m_rt::entry;
+#[macro_use]
+mod utilities;
+use stm32h7xx_hal::hal::digital::v2::{InputPin, OutputPin};
+use stm32h7xx_hal::pwm::{FaultMonitor, Polarity};
+use stm32h7xx_hal::{pac, prelude::*};
+
+use log::info;
+
+#[entry]
+fn main() -> ! {
+    utilities::logger::init();
+    let dp = pac::Peripherals::take().expect("Cannot take peripherals");
+
+    // Constrain and Freeze power
+    info!("Setup PWR...                  ");
+    let pwr = dp.PWR.constrain();
+    let pwrcfg = example_power!(pwr).freeze();
+
+    // Constrain and Freeze clock
+    info!("Setup RCC...                  ");
+    let rcc = dp.RCC.constrain();
+    let ccdr = rcc.sys_ck(8.mhz()).freeze(pwrcfg, &dp.SYSCFG);
```

Comment:
I would specify `.hclk(4.mhz())` directly here. Without this the example breaks on my NUCLEO-H7A3ZI-Q because for this family the HAL defaults to a [sys_clk/hclk ratio of 1 rather than 2](https://github.com/stm32-rs/stm32h7xx-hal/blob/master/src/rcc/mod.rs#L710-L714) and thus ends up with a `timx_ker_ck` of 8MHz rather than the 4MHz you were expecting. That in turn breaks the deadtime on TIM1 (too many ticks) and TIM2 where the period is specified.

You could make the example even clearer with an explicit assert on the clocks you need. Something like:
```rust
assert_eq!(ccdr.clocks.timx_ker_ck().0, 4_000_000);
```
That causes a panic earlier if someone (for example) specifies pclks which push the hclk/timx_ker_clk ratio to something other than 1 (see RM0433 Rev 7 Table 56).